README.md
# ZombireBot
Zombire IRC bot: a multiplayer mini-RPG battle game between zombies and vampires.

## Prerequisites in IRC environment
ZombireBot works properly if run on Rizon IRC network. It would also work on other IRC networks that use Anope services.

## How to Install
1. You need `Python3` and `pip3` installed on your system, and an access to a MySQL database.

2. Run: `pip3 install -r requirements.txt`

3. Register the bot nick on the IRC network you want it to run on, and gives it the ability to change the channel access.

4. Copy `conf/config.example.yml` to `conf/config.yml` and change the configuration parameters according to your needs.

5. Create the mysql database and user, then run (substituting `mysql_user` and `mysql_db` by their values):
```
mysql -u mysql_user -D mysql_db -p < conf/create_tables.sql
```
Finally, run: `python3 zombire.py` and enjoy!

## How to Play
Vampires and zombies are trying to annihilate and destroy each other. Every player joining the game will be either a vampire or a zombie at random. Vampires can attack zombies only, and vice versa. Furthermore, vampires can heal vampires only, and zombies can heal zombies only.

Each player has the following stats:
- **HP :**  those are the health points (they have no maximum limit per player)
- **MP :**  those represent how many remaining actions the player can do in the current hour
- **Maximum MP :**  those are the maximum actions per hour for the player
- **Score :**  score points awarded for healings and successful attacks
- **Bonus :**  if present, it will give a positive or negative boost to the player attack/defense
- **Equipment :**  sword and armor equipped by the player

Each player upon registration will have an HP of 10, and an MP and Maximum MP of 5.

Each player can take 3 types of **actions** against other players:
- **!attack :**  Both players (attacker and target) will throw a dice each. The difference in dice numbers will be subtracted from the loser's HP and added to the winner's HP. In case of a tie, nothing happens. **If a player's HP reaches 0, he/she will transform automatically to the opposite type (vampire -> zombie or zombie -> vampire)**. Players cannot attack players of the same type. This action consumes 1 MP from the attacking player.
- **!heal :**  The player will sacrifice 2 HP to give an ally 1 HP. The player needs at least 3 HP to make this action. Players cannot heal players of the opposite type. This action consumes 1 MP from the healing player.
- **!ambush :**  Player will attack 2 enemies simultaneously. If the ambush succeeds, he/she will gain 6 HP and the other two will loose 3 HP each. If not, he/she will loose 6 HP and the other two gain 3 HP each. The player needs at least 6 HP to make this action, and it will consume 2 MP from his/her stats.

**Every one hour**, the MP of all players will be regenerated and filled up to their respective Maximum MP.

**About Maximum MP**, if a player makes 5 successful attacks/ambushes in a row (with no ties in between), his/her Maximum MP will go up by 1 point. However, if he/she makes 5 failed attacks/ambushes in row, his/her Maximum MP will go down by 1 point.

**Concerning bonus**, every 3 hours, some players may find random items or get affected by random objects, leading to a temporary positive or negative effect on their attack/defense. The bonus effect will last for 1 hour only.

**Scoring:**
- a healing gives 1 score point
- a successful attack gives 2 score points
- a successful ambush gives 3 score points
- attacking and killing an enemy gives 5 additional score points
- successfully attacking and inflicting damage on the enemy leader gives 10 score points (instead of 2)

**Player settings:**
- `!auto register` command, when set to `on`, allows the player to automatically register when a new round of the game begins.
- `!auto heal` command, if set to `lowest`, will make the player automatically heals the ally with the lowest HP, before the current hour ends. If set to `highest`, the target of healing will be the ally with the highest HP. The player should at least 1 MP remaining for that auto-action to take place.
- `!auto attack` command, similarly to `!auto heal`, allows the player to attack the enemy with the lowest or highest HP, before the current hour ends.
- `!auto search` command, will allow the player to auto-spend 1 MP (if available) to search for an item. It won't work of the user inventory is already full.
- `!auto chest` command, will allow the player to automatically open/drop the chests right after obtaining them.
- Note that enabling `!auto attack` will cancel `!auto heal` and `!auto search`, and vice versa.

**Items:**

Players can search for items and use them. Searching for items using the `!search` command will consume 1 MP. The chance of finding an item is 25%. Be aware that the player inventory has a limited capacity of 3 items.

Some of the items are considered **rare items**. A player cannot have more than 1 unit of the same rare item in his/her inventory.

Here is a full list of the items that can be found:

Item | Can be used on others? | Rare? | Description
---- | ---------------------- | ----- | -----------
Small apple | no | no | +2 HP
Medium apple | no | no | +5 HP
Large apple | no | no | +10 HP
Small lemon | no | no | +1 MP (MP will not go over Maximum MP)
Medium lemon | no | no | +2 MP (MP will not go over Maximum MP)
Large lemon | no | no | +3 MP (MP will not go over Maximum MP)
Transformic | no | yes | tranforms the player to the opposite type without affecting his/her stats
Explodic | yes | yes | suicidal attack (renders the HP of the user and the target to 1)
Neutralic | yes | yes | removes any bonus effect of a particular player
Switchic | yes | yes | switches the HP stats between player and target (consumes the full MP of the player)
Drainic | yes | yes | sacrifices 1 MP to decrease the target's HP by 5
Revealic | yes | yes | reveals the inventory contents of the target
Thievic | yes | yes | steals a random item from the target (if the target has any)
Summonic | no | yes | sacrifices 20 HP to summon the player's leader for 15 minutes
Maintainic | no | yes | increases the lifespan of a sword or an armor by 20

The unused items will remain in the player inventory even after a round ends.

**Chests and Equipment:**

Every new player starts with a wooden sword and a wooden armor, both possessing a power of 0, and both upgradable. 

Upon taking an action (attack, ambush or heal), there is a chance that the player will find a chest, which they can open or drop. The chest may contain one of the seven ores: `Copper`, `Steel`, `Titanium`, `Quartz`, `Sapphire`, `Diamond` and `Bomb`. The ores collected are stored in the player's forge, which has a maximum capacity of 4. Except for `Bomb`, which causes up to 5 HP damage to the player, the other ores can be used to upgrade one's sword or armor, using the `!upgrade` command, as follows:

Sword | Ores in forge | Upgraded form
----- | ------------- | -------------
Wooden sword | 3 Copper | Copper sword (power = 1)
Copper sword | 3 Steel | Steel sword (power = 2)
Steel sword | 3 Titanium | Titanium sword (power = 3)

Armor | Ores in forge | Upgraded form
----- | ------------- | -------------
Wooden armor | 3 Quartz | Quartz armor (power = 1)
Quartz armor | 3 Sapphire | Sapphire armor (power = 2)
Sapphire armor | 3 Diamond | Diamond armor (power = 3)

Furthermore, each sword/armor (except for wooden ones) has lifespan that is decreased by 1 after each use of the sword/armor. If the lifespan reaches 0, the sword or armor will be destroyed and the player will get a wooden sword or armor instead. After successful upgrading, the lifespan of a sword/armor is reset to 100. Item `Maintainic` can also be used to increase the lifespan of a sword or an armor by 20.

Swords and armors play an important role in battles (attacks and ambushes). After rolling dices, swords can help to increase the damage and armors to reduce it. The player with the more powerful equipment used will get the favor in damage increase/decrease.

The sword and armor, along with their lifespans, will remain in the player equipment even after a round ends.

**Special apperarences:**

The respective leaders of vampires and zombies, Count Dracula and General Zombilo, will join the battle once a day at random time, for 15 minutes each. When they are around, their soldiers will have a dramatical boost in attack and defense, temporarily removing any other negative effects. When the leader leaves, his soldiers bonus stats return as they were before his arrival. 

Those leaders can be attacked when they appear using the `!challenge` command. Be aware that they use a 12-face dice though, when the player uses a 6-face dice. A successful attack will award the player 10 score points.

**End of Round:**

The game will end when all players become vampires, or all players become zombies. The highscorer player will have his/her name added to the highscore table, and the round will end removing all players from the game. To participate in a new round, the players will have to type the `!register` command again.

### List of user commands
- **!howtoplay :**  displays a link to this page
- **!register :**  transforms te player into a vampire or a zombie and let him/her participate in the game
- **!unregister :**  removes the player from the game
- **!attack _player_ :**  lets the player attack a player of the opposite type (an enemy)
- **!heal _player_ :**  lets the player heal a player of the same type (an ally)
- **!ambush _player1_ _player2_ :**  lets the player attack 2 enemy players simultaneously
- **!status [_player_] :**  displays the stats of the mentioned player, or displays your own stats if you omit _player_
- **!vampires :**  lists all vampires in the current round
- **!zombies :**  lists all zombies in the current round
- **!topscores** or **!highscores :**  displays the top 10 highscores in the game
- **!auto register [off|on] :**  for reading or changing the auto register setting
- **!auto heal [off|lowest|highest] :**  for reading or changing the auto heal setting
- **!auto attack [off|lowest|highest] :**  for reading or changing the auto attack setting
- **!auto search [off|on] :**  for reading or changing the auto search setting
- **!auto chest [off|open|drop] :**  for reading or changing the auto chest setting
- **!challenge :**  attack the enemy leader if he is around
- **!search :**  searches for a random item (add the item to the inventory if found)
- **!inventory :**  lists the items in your inventory
- **!use _number_ :**  uses item in your inventory on yourself (number can be: 1, 2 or 3)
- **!use _number_ _target_ :**  uses item in your inventory on a target (applies only on few certain items)
- **!drop _number_ :**  drops an item from your inventory (number can be: 1, 2 or 3)
- **!chest open|drop :**  for opening a chest or dropping it after finding one
- **!forge :**  lists all the ores in the forge of the player (max. capacity of the forge is 4)
- **!forge drop _number_ :**  drops an ore from the player's forge
- **!upgrade sword|armor :**  upgrades the sword or armor if the necessary ores are present in the player's forge 
- **!version :**  displays the running version of the bot

### List of admin commands
All admin commands are in the following syntax:
```
/msg <botname> admin <admin_passwd> <command> ...
```
where `<botname>` is the registered nick used by the bot, and `<admin_passwd>` is the one configured in `config.yml` file.

The possible commands are:

- **quit _message_ :**  makes the bot disconnect (*message* is optional and will appear in the bot quitting message)
- **kick _player(s)_ :**  removes player(s) from the game (you can specify a list of players separated by spaces)
- **terminate _player(s)_ :**  similar to **kick**, but delete player settings in addition (such as those set by !auto commands)
- **stats :**  displays the total of vampires and zombies in the current round
- **topscores** or **highscores :**  displays the top 10 highscores with more details
- **clearscores :**  deletes all entries from the highscore table


CHANGELOG
CHANGELOG
---------
v1.4.0
- Generate JSOV files for player stats and stuff
- Turn auto-register off for players who !unregister
- Allow chests to appear after !challenge command
- Unregister users when they are kicked/terminated by admin (issue #8)

v1.3.6
- Added the !auto chest command
- Added the possibilty of finding chests after auto attacking/healing
- Fixed some messages

v1.3.5
- Fixed a crash error in !challenge command

v1.3.4
- Fixed a bug in HP not reaching 0 when player has superior armor

v1.3.3
- Fixed a bug in armor upgrade

v1.3.2
- Increased the lifespan of upgraded swords/armors from 40 to 100
- Increased the maximum forge capacity from 3 to 4

v1.3.1
- Fixed an issue with chest not disappering after it opens up on a bomb

v1.3.0
- Do not allow players to have more than 1 unit of each rare item
- Added 4 more items: Revealic, Thievic, Summonic and Maintainic
- Imposed restrictions on the use of item Switchic (it now needs and consumes full MP)
- Made rare items appear half as much as non-rare items
- Added chests and forges to the game (containing ore of metals and jewels)
- Added swords and armors, which affect the damage calculation in battles
- Disabled direct hilighting of nicknames by the bot
- Fixed issue #7 (stats not saving properly when round ends)

v1.2.1
- Added "!auto search" command
- Added 2 more items: Switchic and Drainic

v1.2.0
- Added new user commands for items: !search, !use, !inventory, !drop
- Added ability to fight enemy leaders using: !challenge
- Reward the attacker 5 additional score points if he/she finishes off an enemy
- Allow running !status without arguments to show the status of the issuing player

v1.1.1
- Fixed minor issues in v1.1.0

v1.1.0
- Added the leader characters: Dracula and Zombilo (they give their soldiers substantial boost when they appear)
- Added !auto commands that let the user change its settings to auto register/attack/heal
- Display the round length in time when the round ends
- Added !version command
- Fixed other issues

v1.0.2
- Enabled the round to end when all remaining players of one type unregister

v1.0.1
- Minor fix that prevents bot from crashing when declaring the winner

v1.0.0
- First stable release of Zombire IRC bot


zombirelib/user.py
import random
import datetime

from .utils import Utils


class User:
	CUMULATIVE = 5
	MMP_MAX = 60
	MMP_MIN = 1
	rare_items = (7, 8, 9, 10, 11, 12, 13, 14, 15)
	item_names = {
	0: "",
	1: "small apple",   # +2 HP
	2: "medium apple",  # +5 HP
	3: "large apple",   # +10 HP
	4: "small lemon",   # +1 MP
	5: "medium lemon",  # +2 MP
	6: "large lemon",   # +3 MP
	7: "Tansformic",    # for type transformation
	8: "Explodic",      # for a suicidal attack
	9: "Neutralic",     # for removing bonus effect
	10: "Switchic",     # for switching HP stats with another player
	11: "Drainic",      # sacrifice 1 MP to get 5 HP
	12: "Revealic",     # reveals the inventory contents of a target
	13: "Thievic",      # steals a random item from target (if he has any)
	14: "Summonic",     # sacrifices 20 HP to summon the player's leader
	15: "Maintainic",   # increase sword/armor life by 20
	}
	ore_names = {
	0: "",
	1: "Copper",
	2: "Steel",
	3: "Titanium",
	4: "Quartz",
	5: "Sapphire",
	6: "Diamond",
	7: "Bomb",
	}
	sword_names = {
	0: "Wooden",
	1: "Copper",
	2: "Steel",
	3: "Titanium",
	}
	armor_names = {
	0: "Wooden",
	1: "Quartz",
	2: "Sapphire",
	3: "Diamond",
	}

	@staticmethod
	def is_online(conn, nick):
		conn.whois(nick)

	@staticmethod
	def is_identified(conn, nick):
		conn.privmsg("nickserv", "status {}".format(nick))

	@staticmethod
	def check_eq(players, nick, stats, val):
		return players[nick][stats] == val

	@staticmethod
	def check_gt(players, nick, stats, val):
		return players[nick][stats] > val

	@staticmethod
	def check_lt(players, nick, stats, val):
		return players[nick][stats] < val

	@staticmethod
	def check_ge(players, nick, stats, val):
		return players[nick][stats] >= val

	@staticmethod
	def check_le(players, nick, stats, val):
		return players[nick][stats] <= val

	@staticmethod
	def ratio_of_types(players):
		num_v = 0
		num_z = 0
		num_v = sum(num_v + 1 for nick in players if players[nick]['type'] == "v")
		num_z = sum(num_z + 1 for nick in players if players[nick]['type'] == "z")
		if num_v or num_z:
			return num_v / (num_v + num_z)
		return 0.5

	@staticmethod
	def get_bonus(source, players):
		# lowest 2 bits of 'bonus'
		return players[source]['bonus'] & 3

	@staticmethod
	def set_bonus(source, players, value):
		# value can only be: 0, 1 or 2
		players[source]['bonus'] &= ~3
		players[source]['bonus'] |= value

	@staticmethod
	def reset_both_attacks(source, players):
		# clear bits 2..7 of 'bonus'
		players[source]['bonus'] &= ~(63 << 2)

	@staticmethod
	def check_positive_attacks(source, players):
		# bits 2..4 of 'bonus'
		positive_count = (players[source]['bonus'] >> 2) & 7
		# reset negative count
		players[source]['bonus'] &= ~(7 << 5)
		# increase and check
		positive_count = (positive_count + 1) % User.CUMULATIVE
		# store back
		players[source]['bonus'] &= ~(7 << 2)
		players[source]['bonus'] |= positive_count << 2
		if not positive_count:
			return True

	@staticmethod
	def check_negative_attacks(source, players):
		# bits 5..7 of 'bonus'
		negative_count = (players[source]['bonus'] >> 5) & 7
		# reset positive count
		players[source]['bonus'] &= ~(7 << 2)
		# increase and check
		negative_count = (negative_count + 1) % User.CUMULATIVE
		# store back
		players[source]['bonus'] &= ~(7 << 5)
		players[source]['bonus'] |= negative_count << 5
		if not negative_count:
			return True

	@staticmethod
	def transform(target, players, source=None):
		if players[target]['hp'] > 0:
			return False
		if players[target]['type'] == "v":
			players[target]['type'] = "z"
		else:
			players[target]['type'] = "v"
		players[target]['hp'] = 10
		if source:
			players[source]['score'] += 5
		return True

	@staticmethod
	def donate(source, target, players):
		players[source]['hp'] -= 2
		players[target]['hp'] += 1
		players[source]['mp'] -= 1
		players[source]['score'] += 1

	@staticmethod
	def battle(source, target, players):
		random.seed()
		p1 = players[source]
		p2 = players[target]
		bonus1 = User.get_bonus(source, players)
		bonus2 = User.get_bonus(target, players)
		if bonus1 == 1:
			dice1 = int(random.random() * 4) + 3 # +30% winning chance
		elif bonus1 == 2:
			dice1 = int(random.random() * 4) + 1 # -30% winning chance
		else:
			dice1 = int(random.random() * 6) + 1 # normal winning chance
		if bonus2 == 1:
			dice2 = int(random.random() * 4) + 3 # +30% winning chance
		elif bonus2 == 2:
			dice2 = int(random.random() * 4) + 1 # -30% winning chance
		else:
			dice2 = int(random.random() * 6) + 1 # normal winning chance
		# check for bosses presence
		if Utils.bosses:
			if Utils.bosses[0]['on']:
				if p1['type'] == "v":
					dice1 = random.choice((5, 6))
				else:
					dice2 = random.choice((5, 6))
			if Utils.bosses[1]['on']:
				if p1['type'] == "z":
					dice1 = random.choice((5, 6))
				else:
					dice2 = random.choice((5, 6))
		diff_dice = dice1 - dice2
		players[source]['hp'] += diff_dice
		players[target]['hp'] -= diff_dice
		players[source]['mp'] -= 1
		if diff_dice > 0:
			players[source]['score'] += 2
		if players[source]['hp'] < 0:
			players[target]['hp'] += players[source]['hp']
			if diff_dice > 0:
				diff_dice += players[source]['hp']
			else:
				diff_dice -= players[source]['hp']
			#players[source]['hp'] = 0
		if players[target]['hp'] < 0:
			players[source]['hp'] += players[target]['hp']
			if diff_dice > 0:
				diff_dice += players[target]['hp']
			else:
				diff_dice -= players[target]['hp']
			#players[target]['hp'] = 0
		return [dice1, dice2, diff_dice]

	@staticmethod
	def ambush(source, target1, target2, players):
		random.seed()
		p0 = players[source]
		p1 = players[target1]
		p2 = players[target2]
		bonus0 = User.get_bonus(source, players)
		bonus1 = User.get_bonus(target1, players)
		bonus2 = User.get_bonus(target2, players)
		if bonus0 == 1:
			dice0 = random.randint(3, 6) + random.randint(3, 6)
		elif bonus0 == 2:
			dice0 = random.randint(1, 4) + random.randint(1, 4)
		else:
			dice0 = random.randint(1, 6) + random.randint(1, 6)
		if bonus1 == 1:
			dice1 = random.randint(3, 6)
		elif bonus1 == 2:
			dice1 = random.randint(1, 4)
		else:
			dice1 = random.randint(1, 6)
		if bonus2 == 1:
			dice2 = random.randint(3, 6)
		elif bonus2 == 2:
			dice2 = random.randint(1, 4)
		else:
			dice2 = random.randint(1, 6)
		# check for bosses presence
		if Utils.bosses:
			if Utils.bosses[0]['on']:
				if p0['type'] == "v":
					dice0 = random.choice((5, 6)) + random.choice((5, 6))
				else:
					dice1 = random.choice((5, 6))
					dice2 = random.choice((5, 6))
			if Utils.bosses[1]['on']:
				if p0['type'] == "z":
					dice0 = random.choice((5, 6)) + random.choice((5, 6))
				else:
					dice1 = random.choice((5, 6))
					dice2 = random.choice((5, 6))
		# check for ambush result
		players[source]['mp'] -= 2
		diff_dices = dice0 - dice1 - dice2
		if diff_dices > 0:
			players[source]['score'] += 3
			players[source]['hp'] += 6
			players[target1]['hp'] = max(players[target1]['hp'] - 3, 0)
			players[target2]['hp'] = max(players[target2]['hp'] - 3, 0)
		elif diff_dices < 0:
			players[source]['hp'] -= 6
			players[target1]['hp'] += 3
			players[target2]['hp'] += 3
		return diff_dices

	@staticmethod
	def challenge_boss(source, players):
		random.seed()
		p = players[source]
		bonus1 = User.get_bonus(source, players)
		if p['type'] == "v" and Utils.bosses[0]['on']:
			dice1 = random.choice((5, 6))
		elif p['type'] == "z" and Utils.bosses[1]['on']:
			dice1 = random.choice((5, 6))
		elif bonus1 == 1:
			dice1 = random.randint(3, 6)
		elif bonus1 == 2:
			dice1 = random.randint(1, 4)
		else:
			dice1 = random.randint(1, 6)
		dice2 = random.randint(1, 12)
		diff_dice = dice1 - dice2
		players[source]['mp'] -= 1
		players[source]['hp'] += diff_dice
		if diff_dice > 0:
			players[source]['score'] += 10
		elif diff_dice < 0:
			if players[source]['hp'] < 0:
				diff_dice -= players[source]['hp']
				players[source]['hp'] = 0
		return [diff_dice, dice1, dice2]

	@staticmethod
	def redetermine_mmp(result, nick, players):
		if result > 0:
			if User.check_positive_attacks(nick, players):
				if players[nick]['mmp'] < User.MMP_MAX:
					players[nick]['mmp'] += 1
					return 1
		elif result < 0:
			if User.check_negative_attacks(nick, players):
				if players[nick]['mmp'] > User.MMP_MIN:
					players[nick]['mmp'] -= 1
					return -1
		else:
			# reset the cumulation to zero
			User.reset_both_attacks(nick, players)

	@staticmethod
	def get_inventory(source, profiles):
		# lowest 12 bits of extras, since each item is 4 bits
		items = []
		inventory = profiles[source]['extras'] & 4095
		for i in range(3):
			items.append((inventory >> 4*i) & 15)
		return items

	@staticmethod
	def save_inventory(items, source, profiles):
		inventory = 0
		for i in range(3):
			inventory += items[i] * (16**i)
		profiles[source]['extras'] &= ~4095
		profiles[source]['extras'] |= inventory

	@staticmethod
	def search_item(source, profiles):
		random.seed()
		# 25% chance for finding an item
		chance = random.randint(1, 12)
		if chance == 3:
			# rare item
			return random.randrange(7, len(User.item_names))
		elif chance in (1, 2):
			# non-rare item, aka fruit
			return random.randint(1, 6)
		else:
			return 0

	@staticmethod
	def append_item(new_item, source, profiles):
		items = User.get_inventory(source, profiles)
		if new_item in User.rare_items and new_item in items:
			return False # adding a rare item that already exists isn't possible
		if items[0] and items[1]:
			items[2] = new_item
		elif items[0]:
			items[1] = new_item
		else:
			items[0] = new_item
		User.save_inventory(items, source, profiles)
		return True

	@staticmethod
	def add_item(source, profiles, players):
		items = User.get_inventory(source, profiles)
		if items[0] and items[1] and items[2]:
			return -1  # inventory is full
		players[source]['mp'] -= 1
		new_item = User.search_item(source, profiles)
		if new_item:
			# don't allow a player to have more than 1 unit of a rare item
			while (new_item in User.rare_items and new_item in items) or not new_item:
				new_item = User.search_item(source, profiles)
			if items[0] and items[1]:
				items[2] = new_item
			elif items[0]:
				items[1] = new_item
			else:
				items[0] = new_item
			User.save_inventory(items, source, profiles)
			return new_item

	@staticmethod
	def get_item(item_index, source, profiles):
		return (User.get_inventory(source, profiles))[item_index]

	@staticmethod
	def drop_item(item_index, source, profiles):
		items = User.get_inventory(source, profiles)
		if item_index == 0:
			items[0] = items[1]
			items[1] = items[2]
			items[2] = 0
		elif item_index == 1:
			items[1] = items[2]
			items[2] = 0
		else:
			items[2] = 0
		User.save_inventory(items, source, profiles)

	@staticmethod
	def use_item(item, source, players):
		if item == 1:
			players[source]['hp'] += 2
		elif item == 2:
			players[source]['hp'] += 5
		elif item == 3:
			players[source]['hp'] += 10
		elif item == 4:
			players[source]['mp'] = min(players[source]['mp'] + 1, players[source]['mmp'])
		elif item == 5:
			players[source]['mp'] = min(players[source]['mp'] + 2, players[source]['mmp'])
		elif item == 6:
			players[source]['mp'] = min(players[source]['mp'] + 3, players[source]['mmp'])
		elif item == 7:
			if players[source]['type'] == "v":
				players[source]['type'] = "z"
			else:
				players[source]['type'] = "v"
		elif item == 14:
			players[source]['hp'] -= 20

	@staticmethod
	def use_item2(item, source, target, players):
		if item == 8:
			players[source]['hp'] = 1
			players[target]['hp'] = 1
		elif item == 9:
			User.set_bonus(target, players, 0)
		elif item == 10:
			players[source]['mp'] = 0
			old_hp = players[source]['hp']
			players[source]['hp'] = players[target]['hp']
			players[target]['hp'] = old_hp
		elif item == 11:
			players[source]['mp'] -= 1
			players[target]['hp'] = max(players[target]['hp'] - 5, 0)
		elif item == 12:
			pass # no stats will be affected
		elif item == 13:
			pass # no stats will be affected

	@staticmethod
	def use_item3(item, source, weapon, arsenals):
		if item == 15:
			if weapon == "sword":
				return User.increase_sword_life(source, arsenals, 20)
			elif weapon == "armor":
				return User.increase_armor_life(source, arsenals, 20)

	@staticmethod
	def summon_boss(boss_index):
		if Utils.bosses[boss_index]['on']:
			return False
		now_hour = datetime.datetime.now().hour
		now_min = datetime.datetime.now().minute
		now_sec = datetime.datetime.now().second
		now_min += 1
		if now_min >= 60:
			now_min %= 60
			now_hour = (now_hour + 1) % 24
		Utils.bosses[boss_index]['shown'] = False
		Utils.bosses[boss_index]['h'] = now_hour
		Utils.bosses[boss_index]['m'] = now_min
		Utils.bosses[boss_index]['s'] = now_sec
		return True

	@staticmethod
	def get_forge(source, profiles):
		# bits 13..24 from extras, where each ore is 3 bits
		ores = []
		arsenal = (profiles[source]['extras'] >> 13) & 4095
		for i in range(4):
			ores.append((arsenal >> 3*i) & 7)
		return ores

	@staticmethod
	def save_forge(ores, source, profiles):
		arsenal = 0
		for i in range(4):
			arsenal += ores[i] << 3*i
		profiles[source]['extras'] &= ~(4095 << 13)
		profiles[source]['extras'] |= arsenal << 13

	@staticmethod
	def has_chest(source, profiles):
		# bit 12 in extras
		return (profiles[source]['extras'] >> 12) & 1

	@staticmethod
	def obtain_chest(source, profiles):
		profiles[source]['extras'] |= 1 << 12

	@staticmethod
	def drop_chest(source, profiles):
		profiles[source]['extras'] &= ~(1 << 12)

	@staticmethod
	def open_chest(source, profiles):
		random.seed()
		ore = random.randrange(0, len(User.ore_names))
		#User.drop_chest(source, profiles)
		return ore

	@staticmethod
	def chest_appearing(source, profiles):
		if not User.has_chest(source, profiles):
			random.seed()
			# chance to get a chest is 33%
			chance = random.randint(1, 3)
			if chance == 2:
				User.obtain_chest(source, profiles)
				return True

	@staticmethod
	def add_to_forge(source, players, profiles):
		ores = User.get_forge(source, profiles)
		if ores[0] and ores[1] and ores[2] and ores[3]:
			return -1
		new_ore = User.open_chest(source, profiles)
		if new_ore == 7: # "A BOMB IT IS!" - Yoda
			return new_ore
		if new_ore:
			if ores[0] and ores[1] and ores[2]:
				ores[3] = new_ore
			elif ores[0] and ores[1]:
				ores[2] = new_ore
			elif ores[0]:
				ores[1] = new_ore
			else:
				ores[0] = new_ore
			User.save_forge(ores, source, profiles)
			return new_ore

	@staticmethod
	def drop_from_forge(ore_index, source, profiles):
		# ore_index can be 1, 2, 3 or 4
		ore_index -= 1
		ores = User.get_forge(source, profiles)
		if not ores[ore_index]:
			return False
		if ore_index == 0:
			ores[0] = ores[1]
			ores[1] = ores[2]
			ores[2] = ores[3]
			ores[3] = 0
		elif ore_index == 1:
			ores[1] = ores[2]
			ores[2] = ores[3]
			ores[3] = 0
		elif ore_index == 2:
			ores[2] = ores[3]
			ores[3] = 0
		elif ore_index == 3:
			ores[3] = 0
		User.save_forge(ores, source, profiles)
		return True

	@staticmethod
	def clear_forge(source, profiles):
		User.save_forge([0, 0, 0, 0], source, profiles)

	@staticmethod
	def upgrade_sword(source, profiles, arsenals):
		ores = User.get_forge(source, profiles)
		sword = arsenals[source]['sword']
		nb_of_yes_ores = len([x for x in ores if x == sword + 1])
		if nb_of_yes_ores >= 3:
			arsenals[source]['sword'] += 1
			arsenals[source]['slife'] = 100
			if nb_of_yes_ores == 4:
				User.save_forge([sword + 1, 0, 0, 0], source, profiles)
			else:
				tmp_ores = [x for x in ores if x != sword + 1]
				User.save_forge([(tmp_ores[0]), 0, 0, 0], source, profiles)
			return arsenals[source]['sword']

	@staticmethod
	def upgrade_armor(source, profiles, arsenals):
		ores = User.get_forge(source, profiles)
		armor = arsenals[source]['armor']
		nb_of_yes_ores = len([x for x in ores if x == armor + 4])
		if nb_of_yes_ores >= 3:
			arsenals[source]['armor'] += 1
			arsenals[source]['alife'] = 100
			if nb_of_yes_ores == 4:
				User.save_forge([armor + 4, 0, 0, 0], source, profiles)
			else:
				tmp_ores = [x for x in ores if x != armor + 4]
				User.save_forge([(tmp_ores[0]), 0, 0, 0], source, profiles)
			return arsenals[source]['armor']

	@staticmethod
	def are_weapons_wooden(source, target, arsenals):
		p1 = arsenals[source]
		p2 = arsenals[target]
		or_weapons = p1['sword'] | p1['slife'] | p2['armor'] | p2['alife']
		if not or_weapons:
			return True

	@staticmethod
	def degrade_sword(source, arsenals):
		if arsenals[source]['sword'] > 0:
			arsenals[source]['slife'] -= 1
			if arsenals[source]['slife'] <= 0:
				arsenals[source]['sword'] = 0
				arsenals[source]['slife'] = 0
				return True

	@staticmethod
	def degrade_armor(source, arsenals):
		if arsenals[source]['armor'] > 0:
			arsenals[source]['alife'] -= 1
			if arsenals[source]['alife'] <= 0:
				arsenals[source]['armor'] = 0
				arsenals[source]['alife'] = 0
				return True

	@staticmethod
	def increase_sword_life(source, arsenals, value):
		if arsenals[source]['sword']:
			arsenals[source]['slife'] += value
			return True

	@staticmethod
	def increase_armor_life(source, arsenals, value):
		if arsenals[source]['armor']:
			arsenals[source]['alife'] += value
			return True

	@staticmethod
	def clash_weapons(res, source, target, players, arsenals):
		diff_weapon = 0
		if res > 0:
			diff_weapon = arsenals[source]['sword'] - arsenals[target]['armor']
			if diff_weapon > 0:
				players[source]['hp'] += diff_weapon
				players[target]['hp'] -= diff_weapon
			elif diff_weapon < 0:
				diff_weapon = -min(res, -diff_weapon)
				players[source]['hp'] += diff_weapon
				players[target]['hp'] -= diff_weapon
		elif res < 0:
			diff_weapon = arsenals[target]['sword'] - arsenals[source]['armor']
			if diff_weapon > 0:
				players[target]['hp'] += diff_weapon
				players[source]['hp'] -= diff_weapon
			elif diff_weapon < 0:
				diff_weapon = -min(-res, -diff_weapon)
				players[target]['hp'] += diff_weapon
				players[source]['hp'] -= diff_weapon
		return diff_weapon

	@staticmethod
	def bomb_player(source, players):
		if players[source]['hp'] < 2:
			damage = 0
		elif players[source]['hp'] < 6:
			damage = players[source]['hp'] - 1
		else:
			damage = 5
		players[source]['hp'] -= damage
		return damage

	@staticmethod
	def check_if_round_ended(players):
		len_vamp = len([nick for nick in players if players[nick]['type'] == "v"])
		if len(players) > 0:
			if (not len_vamp) or len_vamp == len(players):
				return max(players, key=lambda nick: players[nick]['score'])

	@staticmethod
	def reset_players(players):
		players.clear()
